src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1244,1818 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+		D_GOTO(out, rc);
+	}
+
+	(*bytes_read) += size;
+	kv_val = &(*hdf5->dk)[*dkey_index].rec_kv_val;
+	D_ALLOC(kv_val->p, (uint64_t)size);
+	if (kv_val->p == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	memcpy(kv_val->p, buf, (uint64_t)size);
+	kv_val->len = (uint64_t)size; 
+out:
+	D_FREE(buf);
+	return rc;
+}
+
+static int
+serialize_dkeys(struct hdf5_args *hdf5, daos_obj_id_t oid, daos_handle_t coh, int *oid_index,
+		int *dkey_index, int *akey_index, daos_handle_t *oh, bool is_kv,
+		int *total_dkeys, int *total_akeys, uint64_t *bytes_read)
+{
+	int rc = 0;
+	int		i = 0;
+	daos_anchor_t	dkey_anchor = {0}; 
+	d_sg_list_t	dkey_sgl;
+	d_iov_t		dkey_iov;
+	daos_key_desc_t	dkey_kds[ENUM_DESC_NR] = {0};
+	uint32_t	dkey_number = ENUM_DESC_NR;
+	char		*dkey = NULL;
+	char		*dkey_ptr;
+	daos_key_t	diov;
+	int		path_len = 0;
+	hvl_t		*dkey_val;
+	char		key_val[ENUM_KEY_BUF];
+	char		*small_key = NULL;
+	char		*large_key = NULL;
+	char		*key_buf = NULL;
+	daos_size_t	key_buf_len = 0;
+
+	D_ALLOC(small_key, ENUM_DESC_BUF);
+	if (small_key == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	D_ALLOC(large_key, ENUM_LARGE_KEY_BUF);
+	if (large_key == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	(*hdf5->oid)[*oid_index].dkey_offset = *dkey_index;
+	while (!daos_anchor_is_eof(&dkey_anchor)) {
+		memset(dkey_kds, 0, sizeof(dkey_kds));
+		memset(small_key, 0, ENUM_DESC_BUF);
+		memset(large_key, 0, ENUM_LARGE_KEY_BUF);
+
+		dkey_number 	   = ENUM_DESC_NR;
+		dkey_sgl.sg_nr     = 1;
+		dkey_sgl.sg_nr_out = 0;
+		dkey_sgl.sg_iovs   = &dkey_iov;
+
+		key_buf = small_key;
+		key_buf_len = ENUM_DESC_BUF;
+		d_iov_set(&dkey_iov, key_buf, key_buf_len);
+
+		if (is_kv) {
+			/* TODO: daos_kv_list in src/include/daos_kv.h does not
+			 * document that it returns DER_KEY2BIG so it just uses
+			 * the small buffer. If it does return DER_KEY2BIG but
+			 * it was not documented this will need to be updated. */
+			rc = daos_kv_list(*oh, DAOS_TX_NONE, &dkey_number, dkey_kds,
+					  &dkey_sgl, &dkey_anchor, NULL);
+			if (rc != 0) {
+				D_ERROR("failed to list dkeys "DF_RC"\n", DP_RC(rc));
+				D_GOTO(out, rc);
+			}
+		} else {
+			rc = daos_obj_list_dkey(*oh, DAOS_TX_NONE, &dkey_number,
+						dkey_kds, &dkey_sgl, &dkey_anchor, NULL);
+			if (rc == -DER_KEY2BIG) {
+				/* call list dkey again with bigger buffer */
+				key_buf = large_key;
+				key_buf_len = ENUM_LARGE_KEY_BUF;
+				d_iov_set(&dkey_iov, key_buf, key_buf_len);
+				rc = daos_obj_list_dkey(*oh, DAOS_TX_NONE, &dkey_number, dkey_kds,
+							&dkey_sgl, &dkey_anchor, NULL);
+				if (rc != 0) {
+					D_ERROR("failed to list dkeys "DF_RC"\n", DP_RC(rc));
+					D_GOTO(out, rc);
+				}
+			}
+		}
+
+		/* if no dkeys were returned move on */
+		if (dkey_number == 0)
+			continue;
+
+		*hdf5->dk = realloc(*hdf5->dk, (dkey_number + *total_dkeys) * sizeof(struct dkey));
+		if (*hdf5->dk == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+		
+		/* parse out individual dkeys based on key length and
+		 * number of dkeys returned
+ 		 */
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs